### ISSUE TITLE
Java 8 support for client libraries

### ISSUE THREAD

**Issue by avgustinmm** [eclipse/hono#1137]
From Hono 1.0-M1 hono requires Java 11 because of Java 9 issues.
Are these Java 8 problems related to hono-client and its dependencies (hono-core)?
If not, maybe it'll be better to keep them Java 8 compatible.
It's probably ok to force hono backend side to java 11, But do you feel it's ok to force client side (e.g. backend application) to switch to Java 11?

**Comment by ctron** [comment]
The root cause for that is tracked in: https://github.com/eclipse/hono/issues/932

Yes, it was the client module. And it was in the test source.

However I don't think it is a good idea to compile the main artifact with Java 8, and then compile the test source with Java 11. I also don't think that we should start switching partially to Java 8, as this will immediately introduce the issue of needing different Java toolchains for validating that is still compiles in older versions. As you need to validate against the specific Java API.

I know with OSGi, execution environment, Tycho, this is all not too hard. But for plain Maven, we would need to start setting up each environment (Travis, Jenkins, local machine) with two JDKs.

So I think we should stick to Java 11 overall. Unless there are some reasons that make it absolutely necessary.

**Comment by sophokles73** [comment]
The problem we ran into with #932 was not so much an issue of the source/target level but of the JDK being used to compile and run the tests. From a user perspective, I can sympathize with @avgustinmm 's intent to keep compatibility with Java 8 target level because many applications will remain to be built using Java 8 for some time. In fact, Amazon and RedHat are supporting this explicitly by means of continuing to release and support Java 8 level JDKs.
I can imagine simply defining source/target level 1.8 for the core and client modules, thus making sure the resulting artifacts can be used on Java 8 JVMs. We might still be able to configure the Maven build to require a Java 11 JDK in order to mitigate the problem described in #932 ...
I don't think that it would be an issue to have the core/client modules at Java level 1.8 and have the other artifacts at Java 11 level. I also do not see any reason why this would imply having to set up multiple tool chains. FMPOV we could stick with our existing tool chain which uses OpenJDK 11 only.

**Comment by ctron** [comment]
> I don't think that it would be an issue to have the core/client modules at Java level 1.8 and have the other artifacts at Java 11 level.

What can easily happen is that people use functionality that is not present in the Java 8 APIs. That would result in a Java 8 compiled binary, which fails when being run on Java 8. So the goal of supporting Java 8 would not be accomplished.

I ran into this a few time, typical example was `String.isEmpty()`, which was introduced in Java 6, but people compiled this with Java 6, for Java 5. Resulting in something that passed all tests, and later failed to run in Java 5.

**Comment by sophokles73** [comment]
> What can easily happen is that people use functionality that is not present in the Java 8 APIs

This can be prevented by using the *Animal Sniffer* Maven plugin.
I will prepare a corresponding PR. I think it would be worth the effort to keep Java 8 compatibility. After all, we are not using/requiring any Java 11 language features in the core and client modules anyway ...

**Comment by avgustinmm** [comment]
I am completely agree, that if it would be easy to keep Java 8 compatibility for the client libraries then it should. Java 8 is still here and it will be for some time. So, if there is no technical reason to drop support - better to keep it.

**Comment by sophokles73** [comment]
@avgustinmm is it possible for you to check out #1140 and provide some feedback in the PR whether this works for you?

**Comment by avgustinmm** [comment]
@sophokles73,
I built the PR and tried to install bundles on OSGi & Java 8. I have problems to install them because of manifest header (in both bundles):
  Require-Capability: osgi.ee;filter:="(osgi.ee=UNKNOWN)"
It may be because of problem with my setup - not sure. However if I fix that header the bundles are installed and started successfully.

**Comment by sophokles73** [comment]
@avgustinmm 
Hm, I am no OSGi expert. In fact, I have no idea how I can influence the value of this header in the corresponding Maven POM. Do you have any idea what the value should actually be?

**Comment by ctron** [comment]
I just ran this locally and ended up with (for both client and core):

~~~
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
~~~

**Comment by sophokles73** [comment]
Yes, it works for me as well. Maybe @avgustinmm can try again to do a full mvn clean install on Hono and see if it works for him as well?

**Comment by avgustinmm** [comment]
I've built it with clean build but with OpenJDK (I think this might be  important). I just checked https://search.maven.org/artifact/org.eclipse.hono/hono-client/1.0-M1/bundle - the release artifact. It also has: 
Require-Capability: osgi.ee;filter:="(osgi.ee=UNKNOWN)"
while  https://search.maven.org/artifact/org.eclipse.hono/hono-client/0.9/bundle has:
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
So, the "unknown" problem is not only with the branch but with 1.0 development. Maybe with change of the build JDK.

**Comment by sophokles73** [comment]
@avgustinmm what JDK are you using for building?

**Comment by avgustinmm** [comment]
We use:
openjdk version "11.0.2" 2019-01-15
OpenJDK Runtime Environment 18.9 (build 11.0.2+9)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+9, mixed mode)
btw, we tested to update maven-bundle-plugin (in root pom.xml) to 4.1.0
now we have: Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=11))"
which is an improvement (probably old plugins just don't know for newest jdks). This shall be made in all cases (nevertheless you add support for java8 or not). 
But if you add 8 support the requirement shall be something like osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
We tested with adding in core pom.xml in maven-bundl-plugin instructions:
<Require-Capability>osgi.ee;filter:="(&amp;(osgi.ee=JavaSE)(version=8))"</Require-Capability>
with that all is ok.
So, I'd propose to add this for both core and client. This shall fix the problem for us. And probably to update version of plugin

**Comment by sophokles73** [comment]
Here's what I find in core module's `MANIFEST.MF` when building the #1140 branch using bundle plugin 4.1.0:
```
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Tool: Bnd-4.1.0.201810181252
```
Here's my JDK version:
```
openjdk version "11.0.1" 2018-10-16
OpenJDK Runtime Environment (build 11.0.1+13-Ubuntu-3ubuntu3.18.10.1)
OpenJDK 64-Bit Server VM (build 11.0.1+13-Ubuntu-3ubuntu3.18.10.1, mixed mode, sharing)
```
However, I agree that it's probably best to explicitly define the required capability in the POM.

**Comment by avgustinmm** [comment]
We found out that had problems with build .. the initial PR probably would build correct requirements.
Anyway the last version is ok for us.
